doc/website: add usage document for the view feature in GORM provider (#2797)
ariga/atlas: doc/md/guides/orms/gorm.md

### Standalone vs Go Program mode

The Atlas GORM Provider can be used in two modes:
* **Standalone** - If all of your GORM models and views exist in a single package, and the models either embed `gorm.Model` or contain `gorm` struct tags, 
  you can use the provider directly to load your GORM schema into Atlas.
* **Go Program** - If your GORM models or views are spread across multiple packages, or the models do not embed `gorm.Model` or contain `gorm` struct tags, 
  you can use the provider as a library in your Go program to load your GORM schema into Atlas.

### Standalone mode

> If all of your GORM models and [view definition](https://github.com/ariga/atlas-provider-gorm?tab=readme-ov-file#views) exist in a single package, and the models either embed `gorm.Model` or contain `gorm` struct tags,

Reviewer: No need to mention it in every sentence. Models can be presented as views as well.

Reviewer 2: Agree. Leave it on L55 and in the rest refers to it as "models"

-- Modify "pets" table
ALTER TABLE `pets` ADD COLUMN `nickname` longtext NULL;
```

> #### View

Reviewer: cc @rotemtam for review

> This is a unique API for the Atlas GORM Provider that allows you to define database views in the form of GORM models and automatically generates migration files for you.

Reviewer 2: ```suggestion
From a querying perspective, views and tables are identical. For this reason, GORM can natively query any views that exist on the database.
However, defining and managing views has previously had [partial support](https://github.com/go-gorm/gorm/issues/4966).
The Atlas GORM Provider provides an API that allows you to define database views in the form of GORM models and automatically generate migrations to manage them.
```

> The view feature is only available for logged-in users; run `atlas login` if you haven't already.

> To define a Go struct as a database `VIEW`, implement the `ViewDef` method as follows:

Reviewer 2: ```suggestion
To define a Go struct as a database `VIEW`, implement the [`ViewDefiner`](https://pkg.go.dev/ariga.io/atlas-provider-gorm/gormschema#ViewDefiner) interface:
```

add the interface definition and explain about `ViewOption`

> ```go

Reviewer 2: Views are essentially saved queries, so to define them we must define a query. 

The `gormschema` package provide two styles for defining a view's base query:
* BuildStmt - explain what it is and when to use
* CreateStmt - explain what it is and when to use.
Then give a short example for each.

-- Modify "pets" table
ALTER TABLE `pets` ADD COLUMN `nickname` longtext NULL;
```
### View

> A database view is a virtual table based on the result of a query. Views are helpful when you want to simplify complex queries, strengthen security by selecting only the necessary data, or encapsulate the details of your table structures.

Reviewer 3: ```suggestion

A database view is a virtual table based on the result of a query. Views are helpful when you want to simplify complex queries, strengthen security by selecting only the necessary data, or encapsulate the details of your table structures.
```

> Making liberal use of views is a key aspect of good SQL database design. Views allow you to encapsulate the details of your table structures, which might change as your application evolves, behind consistent interfaces.
> 
> — <cite>**Postgres documentation**</cite>

From a querying perspective, views and tables are identical. For this reason, GORM can natively query any views that exist on the database.
However, defining and managing views has previously had [partial support](https://github.com/go-gorm/gorm/issues/4966).
The Atlas GORM Provider provides an API that allows you to define database views in the form of GORM models and with the help of [Atlas](https://atlasgo.io/getting-started), migration files can be automatically generated for them.

> > The view feature is only available for logged-in users; run `atlas login` if you haven't already.

Reviewer 3: ```suggestion

> The view feature is only available for logged-in users; run `atlas login` if you haven't already.
```

-- Modify "pets" table
ALTER TABLE `pets` ADD COLUMN `nickname` longtext NULL;
```
### View

A database view is a virtual table based on the result of a query. Views are helpful when you want to simplify complex queries, strengthen security by selecting only the necessary data, or encapsulate the details of your table structures.

> Making liberal use of views is a key aspect of good SQL database design. Views allow you to encapsulate the details of your table structures, which might change as your application evolves, behind consistent interfaces.
> 

> > — <cite>**Postgres documentation**</cite>

Reviewer 2: This is great for a blog post. Less so in "dry" tech docs. Keep it in your "stash" for your upcoming post :-)

To define a Go struct as a database `VIEW`, implement the [`ViewDefiner`](https://pkg.go.dev/ariga.io/atlas-provider-gorm/gormschema#ViewDefiner) interface. The `gormschema` package provide two styles for defining a view's base query:

##### BuildStmt

The `BuildStmt` function allows you to define a query using the GORM API. This is useful when you need to use GORM's query building capabilities.
```go
package models

import "ariga.io/atlas-provider-gorm/gormschema"

// WorkingAgedUsers is mapped to the VIEW definition below.
type WorkingAgedUsers struct {
  Name string
  Age  int
}

func (WorkingAgedUsers) ViewDef(dialect string) []gormschema.ViewOption {
  return []gormschema.ViewOption{
    gormschema.BuildStmt(func(db *gorm.DB) *gorm.DB {
      return db.Model(&User{}).Where("age BETWEEN 18 AND 65").Select("name, age")
    }),
  }
}
```
##### CreateStmt

The `CreateStmt` function allows you to define a query using raw SQL. This is useful when you need to use SQL features that GORM does not support.
```go
func (WorkingAgedUsers) ViewDef(dialect string) []gormschema.ViewOption {
  return []gormschema.ViewOption{

>     gormschema.CreateStmt("CREATE VIEW working_aged_users AS SELECT name, age FROM users WHERE age BETWEEN 18 AND 65"),

Reviewer 2: put a comment + example about dialect specific views (SQL syntax varies as you shown in the repo readme)

  }
}
```
The View feature works in both **Standalone** and **Go Program** mode. 

For **Standalone** mode, if you place the view definition in the same package as your models, the provider will automatically detect and create migration files for them.

For **Go Program** mode, you can load the `VIEW` definition in the same way as a model:
```go
gormschema.New("mysql").Load(
  &models.User{},             // Table-based model.
  &models.WorkingAgedUsers{}, // View-based model.
)
```

> With the atlas.hcl configuration file set up as before, you now can run `atlas migrate diff --env gorm` to generate the migration file. For more information and usage of `VIEW`, please refer to the documentation [Atlas GORM Provider#View](https://github.com/ariga/atlas-provider-gorm?tab=readme-ov-file#views)

Reviewer 2: we typically limit lines to ~120 chars, becomes annoying to work with the editor otherwise :-D

##### BuildStmt

The `BuildStmt` function allows you to define a query using the GORM API. This is useful when you need to use GORM's query building capabilities.

```go
package models

import "ariga.io/atlas-provider-gorm/gormschema"

// WorkingAgedUsers is mapped to the VIEW definition below.
type WorkingAgedUsers struct {
  Name string
  Age  int
}

func (WorkingAgedUsers) ViewDef(dialect string) []gormschema.ViewOption {
  return []gormschema.ViewOption{
    gormschema.BuildStmt(func(db *gorm.DB) *gorm.DB {
      return db.Model(&User{}).Where("age BETWEEN 18 AND 65").Select("name, age")
    }),
  }
}
```

##### CreateStmt

The `CreateStmt` function allows you to define a query using raw SQL. This is useful when you need to use SQL features that GORM does not support.

```go
func (WorkingAgedUsers) ViewDef(dialect string) []gormschema.ViewOption {
  var stmt string
  switch dialect {
  case "mysql":
    stmt = "CREATE VIEW working_aged_users AS SELECT name, age FROM users WHERE age BETWEEN 18 AND 65"
  case "sqlserver":

>     stmt = "CREATE VIEW working_aged_users AS SELECT name, age FROM users WHERE age >= 18 AND age <= 65"

Reviewer 3: ```suggestion
    stmt = "CREATE VIEW [working_aged_users] WITH SCHEMABINDING AS SELECT [name], [age] FROM [users] WHERE [age] >= 18 AND [age] <= 65"
```

>     stmt = "CREATE VIEW [working_aged_users] ( [name], [age] ) AS SELECT [name], [age] FROM [users] WHERE [age] >= 18 AND [age] <= 65 WITH CHECK OPTION"

Reviewer 3: ```suggestion
    stmt = "CREATE VIEW [working_aged_users] ([name], [age]) AS SELECT [name], [age] FROM [users] WHERE [age] >= 18 AND [age] <= 65 WITH CHECK OPTION"
```
